src/i_flmusic.c

```diff
@@ -344,6 +345,8 @@ static boolean I_FL_OpenStream(void *data, ALsizei size, ALenum *format,
     if (IsMid(data, size))
     {
         result = fluid_player_add_mem(player, data, size);
+
+        S_SetMusicLumpFormatStr("MIDI (FluidSynth)");
```

Comment:
I suggest adding new function to device interface.

<details>

```diff
diff --git a/src/i_flmusic.c b/src/i_flmusic.c
index 962a400c..64fcf5f0 100644
--- a/src/i_flmusic.c
+++ b/src/i_flmusic.c
@@ -43,7 +43,6 @@ typedef fluid_long_long_t fluid_int_t;
 #include "m_misc.h"
 #include "memio.h"
 #include "mus2mid.h"
-#include "s_sound.h"
 #include "w_wad.h"
 #include "z_zone.h"
 
@@ -345,8 +344,6 @@ static boolean I_FL_OpenStream(void *data, ALsizei size, ALenum *format,
     if (IsMid(data, size))
     {
         result = fluid_player_add_mem(player, data, size);
-
-        S_SetMusicLumpFormatStr("MIDI (FluidSynth)");
     }
     else
     {
@@ -367,8 +364,6 @@ static boolean I_FL_OpenStream(void *data, ALsizei size, ALenum *format,
 
         mem_fclose(instream);
         mem_fclose(outstream);
-
-        S_SetMusicLumpFormatStr("MUS (FluidSynth)");
     }
 
     if (result != FLUID_OK)
@@ -486,6 +481,11 @@ static void I_FL_BindVariables(void)
     BIND_BOOL_MIDI(mus_reverb, false, "FluidSynth reverb");
 }
 
+static const char *I_FL_MusicFormat(void)
+{
+    return "MIDI (FluidSynth)";
+}
+
 stream_module_t stream_fl_module =
 {
     I_FL_InitStream,
@@ -496,4 +496,5 @@ stream_module_t stream_fl_module =
     I_FL_ShutdownStream,
     I_FL_DeviceList,
     I_FL_BindVariables,
+    I_FL_MusicFormat,
 };
diff --git a/src/i_midimusic.c b/src/i_midimusic.c
index 3b8e1d6b..23a94e7b 100644
--- a/src/i_midimusic.c
+++ b/src/i_midimusic.c
@@ -31,7 +31,6 @@
 #include "midifallback.h"
 #include "midifile.h"
 #include "mus2mid.h"
-#include "s_sound.h"
 
 static SDL_Thread *player_thread_handle;
 static SDL_mutex *music_lock;
@@ -1133,8 +1132,6 @@ static boolean RegisterSong(void)
     if (IsMid(song.lump_data, song.lump_length))
     {
         song.file = MIDI_LoadFile(song.lump_data, song.lump_length);
-
-        S_SetMusicLumpFormatStr("MIDI (Native MIDI)");
     }
     else
     {
@@ -1159,8 +1156,6 @@ static boolean RegisterSong(void)
 
         mem_fclose(instream);
         mem_fclose(outstream);
-
-        S_SetMusicLumpFormatStr("MUS (Native MIDI)");
     }
 
     if (song.file == NULL)
@@ -1498,6 +1493,11 @@ static midiplayertype_t I_MID_MidiPlayerType(void)
     return midiplayer_native;
 }
 
+static const char *I_MID_MusicFormat(void)
+{
+    return "MIDI (Native)";
+}
+
 static void I_MID_BindVariables(void)
 {
     BIND_NUM_MIDI(midi_complevel, COMP_STANDARD, 0, COMP_NUM - 1,
@@ -1525,4 +1525,5 @@ music_module_t music_mid_module =
     I_MID_DeviceList,
     I_MID_BindVariables,
     I_MID_MidiPlayerType,
+    I_MID_MusicFormat,
 };
diff --git a/src/i_oalmusic.c b/src/i_oalmusic.c
index 1c848c24..a52548f8 100644
--- a/src/i_oalmusic.c
+++ b/src/i_oalmusic.c
@@ -486,6 +486,15 @@ static void I_OAL_BindVariables(void)
     }
 }
 
+static const char *I_OAL_MusicFormat(void)
+{
+    if (active_module)
+    {
+        return active_module->I_MusicFormat();
+    }
+    return "None";
+}
+
 music_module_t music_oal_module =
 {
     I_OAL_InitMusic,
@@ -500,4 +509,5 @@ music_module_t music_oal_module =
     I_OAL_DeviceList,
     I_OAL_BindVariables,
     I_OAL_MidiPlayerType,
+    I_OAL_MusicFormat,
 };
diff --git a/src/i_oalstream.h b/src/i_oalstream.h
index 4e3d9a59..07204ce4 100644
--- a/src/i_oalstream.h
+++ b/src/i_oalstream.h
@@ -32,6 +32,7 @@ typedef struct
     void (*I_ShutdownStream)(void);
     const char **(*I_DeviceList)(void);
     void (*BindVariables)(void);
+    const char *(*I_MusicFormat)(void);
 } stream_module_t;
 
 extern stream_module_t stream_opl_module;
diff --git a/src/i_oplmusic.c b/src/i_oplmusic.c
index 47854a73..c3576585 100644
--- a/src/i_oplmusic.c
+++ b/src/i_oplmusic.c
@@ -1535,8 +1535,6 @@ static boolean I_OPL_OpenStream(void *data, ALsizei size, ALenum *format,
     if (IsMid(data, size) /* && size < MAXMIDLENGTH */)
     {
         midifile = MIDI_LoadFile(data, size);
-
-        S_SetMusicLumpFormatStr("MIDI (OPL)");
     }
     else
     {
@@ -1561,8 +1559,6 @@ static boolean I_OPL_OpenStream(void *data, ALsizei size, ALenum *format,
 
         mem_fclose(instream);
         mem_fclose(outstream);
-
-        S_SetMusicLumpFormatStr("MUS (OPL)");
     }
 
     if (midifile == NULL)
@@ -1690,6 +1686,11 @@ static const char **I_OPL_DeviceList(void)
     return devices;
 }
 
+static const char *I_OPL_MusicFormat(void)
+{
+    return "MIDI (OPL)";
+}
+
 static void I_OPL_BindVariables(void)
 {
     BIND_NUM_MIDI(num_opl_chips, 1, 1, OPL_MAX_CHIPS,
@@ -1708,4 +1709,5 @@ stream_module_t stream_opl_module =
     I_OPL_ShutdownStream,
     I_OPL_DeviceList,
     I_OPL_BindVariables,
+    I_OPL_MusicFormat,
 };
diff --git a/src/i_sndfile.c b/src/i_sndfile.c
index 30069dd5..b646a191 100644
--- a/src/i_sndfile.c
+++ b/src/i_sndfile.c
@@ -678,20 +678,6 @@ static boolean I_SND_OpenStream(void *data, ALsizei size, ALenum *format,
     *freq = stream.sfinfo.samplerate;
     *frame_size = stream.frame_size;
 
-    int count;
-    sf_command(NULL, SFC_GET_SIMPLE_FORMAT_COUNT, &count, sizeof(count));
-
-    for (int k = 0; k < count; k++)
-    {
-      SF_FORMAT_INFO format_info = {.format = k};
-      sf_command(NULL, SFC_GET_SIMPLE_FORMAT, &format_info, sizeof(format_info));
-      if (format_info.format == stream.sfinfo.format)
-      {
-        S_SetMusicLumpFormatStr(format_info.name);
-        break;
-      }
-    }
-
     return true;
 }
 
@@ -753,6 +739,28 @@ static void I_SND_BindVariables(void)
     ;
 }
 
+static const char *I_SND_MusicFormat(void)
+{
+    static SF_FORMAT_INFO format_info;
+
+    int count;
+    sf_command(NULL, SFC_GET_SIMPLE_FORMAT_COUNT, &count, sizeof(count));
+
+    for (int i = 0; i < count; i++)
+    {
+        format_info.format = i;
+        sf_command(NULL, SFC_GET_SIMPLE_FORMAT, &format_info,
+                   sizeof(format_info));
+        if (format_info.format == stream.sfinfo.format)
+        {
+            return format_info.name;
+            break;
+        }
+    }
+
+    return "Unknown";
+}
+
 stream_module_t stream_snd_module =
 {
     I_SND_InitStream,
@@ -763,4 +771,5 @@ stream_module_t stream_snd_module =
     I_SND_ShutdownStream,
     I_SND_DeviceList,
     I_SND_BindVariables,
+    I_SND_MusicFormat,
 };
diff --git a/src/i_sound.c b/src/i_sound.c
index 551699ee..5cfac311 100644
--- a/src/i_sound.c
+++ b/src/i_sound.c
@@ -666,8 +666,6 @@ boolean IsMus(byte *mem, int len)
 
 void *I_RegisterSong(void *data, int size)
 {
-    S_SetMusicLumpFormatStr("unknown");
-
     for (int i = 0; i < arrlen(music_modules); ++i)
     {
         void *result = music_modules[i]->I_RegisterSong(data, size);
@@ -728,6 +726,15 @@ const char **I_DeviceList(void)
     return devices;
 }
 
+const char *I_MusicFormat(void)
+{
+    if (active_module)
+    {
+        return active_module->I_MusicFormat();
+    }
+    return "Unknown";
+}
+
 void I_BindSoundVariables(void)
 {
     M_BindNum("sfx_volume", &snd_SfxVolume, NULL, 8, 0, 15, ss_none, wad_no,
diff --git a/src/i_sound.h b/src/i_sound.h
index 49e92b7f..b1de0da5 100644
--- a/src/i_sound.h
+++ b/src/i_sound.h
@@ -167,6 +167,7 @@ typedef struct
     const char **(*I_DeviceList)(void);
     void (*I_BindVariables)(void);
     midiplayertype_t (*I_MidiPlayerType)(void);
+    const char *(*I_MusicFormat)(void);
 } music_module_t;
 
 extern music_module_t music_oal_module;
@@ -207,6 +208,8 @@ boolean IsMid(byte *mem, int len);
 // Determine whether memory block is a .mus file
 boolean IsMus(byte *mem, int len);
 
+const char *I_MusicFormat(void);
+
 void I_BindSoundVariables(void);
 
 #endif
diff --git a/src/i_xmp.c b/src/i_xmp.c
index a9499a4d..199bbe9a 100644
--- a/src/i_xmp.c
+++ b/src/i_xmp.c
@@ -20,7 +20,6 @@
 #include "i_oalstream.h"
 #include "i_printf.h"
 #include "i_sound.h"
-#include "s_sound.h"
 
 static xmp_context context;
 
@@ -96,10 +95,6 @@ static boolean I_XMP_OpenStream(void *data, ALsizei size, ALenum *format,
     *freq = SND_SAMPLERATE;
     *frame_size = 2 * sizeof(short);
 
-    struct xmp_module_info info;
-    xmp_get_module_info(context, &info);
-    S_SetMusicLumpFormatStr(info.mod->type);
-
     return true;
 }
 
@@ -160,6 +155,18 @@ static void I_XMP_BindVariables(void)
     ;
 }
 
+static const char *I_XMP_MusicFormat(void)
+{
+    if (!context)
+    {
+        return "Unknown";
+    }
+
+    static struct xmp_module_info info;
+    xmp_get_module_info(context, &info);
+    return info.mod->type;
+}
+
 stream_module_t stream_xmp_module =
 {
     I_XMP_InitStream,
@@ -170,4 +177,5 @@ stream_module_t stream_xmp_module =
     I_XMP_ShutdownStream,
     I_XMP_DeviceList,
     I_XMP_BindVariables,
+    I_XMP_MusicFormat,
 };
diff --git a/src/s_sound.c b/src/s_sound.c
index d1a9c73a..69e0cf42 100644
--- a/src/s_sound.c
+++ b/src/s_sound.c
@@ -634,16 +634,6 @@ void S_SetSfxVolume(int volume)
 }
 
 static int current_musicnum = -1;
-static char music_lump_format_str[40];
-
-void S_SetMusicLumpFormatStr(const char *str)
-{
-    if (str)
-    {
-        M_StringCopy(music_lump_format_str, str,
-                     sizeof(music_lump_format_str));
-    }
-}
 
 void S_ChangeMusic(int musicnum, int looping)
 {
@@ -695,7 +685,7 @@ void S_ChangeMusic(int musicnum, int looping)
     I_Printf(VB_DEBUG, "S_ChangeMusic: %.8s (%s), %s",
              lumpinfo[music->lumpnum].name,
              W_WadNameForLump(music->lumpnum),
-             music_lump_format_str);
+             I_MusicFormat());
 
     mus_playing = music;
 
@@ -744,7 +734,7 @@ void S_ChangeMusInfoMusic(int lumpnum, int looping)
     I_Printf(VB_DEBUG, "S_ChangeMusInfoMusic: %.8s (%s), %s",
              lumpinfo[music->lumpnum].name,
              W_WadNameForLump(music->lumpnum),
-             music_lump_format_str);
+             I_MusicFormat());
 
     mus_playing = music;
 
diff --git a/src/s_sound.h b/src/s_sound.h
index 3393c201..167fa676 100644
--- a/src/s_sound.h
+++ b/src/s_sound.h
@@ -97,8 +97,6 @@ void S_StartMusic(int music_id);
 void S_ChangeMusic(int music_id, int looping);
 void S_ChangeMusInfoMusic(int lumpnum, int looping);
 
-void S_SetMusicLumpFormatStr(const char *str);
-
 // Stops the music fer sure.
 void S_StopMusic(void);
 
```
</details>

Reply:
Thank you for the patch! However, this doesn't distinguish between "MIDI" and "MUS" anymore.

Reply:
> However, this doesn't distinguish between "MIDI" and "MUS" anymore.

Do we need to? Native MIDI module register the song in the thread, so it will complicate things a bit.

Reply:
Fixed patch:
<details>

```diff
diff --git a/src/i_flmusic.c b/src/i_flmusic.c
index 962a400c..7964ebef 100644
--- a/src/i_flmusic.c
+++ b/src/i_flmusic.c
@@ -23,6 +23,7 @@
 #include "config.h"
 #include "i_oalstream.h"
 #include "m_config.h"
+#include "s_musinfo.h"
 
 #if (FLUIDSYNTH_VERSION_MAJOR < 2 \
      || (FLUIDSYNTH_VERSION_MAJOR == 2 && FLUIDSYNTH_VERSION_MINOR < 2))
@@ -43,7 +44,6 @@ typedef fluid_long_long_t fluid_int_t;
 #include "m_misc.h"
 #include "memio.h"
 #include "mus2mid.h"
-#include "s_sound.h"
 #include "w_wad.h"
 #include "z_zone.h"
 
@@ -318,6 +318,8 @@ static boolean I_FL_InitStream(int device)
     return true;
 }
 
+static const char *music_format = "Unknown";
+
 static boolean I_FL_OpenStream(void *data, ALsizei size, ALenum *format,
                                ALsizei *freq, ALsizei *frame_size)
 {
@@ -345,8 +347,7 @@ static boolean I_FL_OpenStream(void *data, ALsizei size, ALenum *format,
     if (IsMid(data, size))
     {
         result = fluid_player_add_mem(player, data, size);
-
-        S_SetMusicLumpFormatStr("MIDI (FluidSynth)");
+        music_format = "MIDI (FluidSynth)";
     }
     else
     {
@@ -367,8 +368,7 @@ static boolean I_FL_OpenStream(void *data, ALsizei size, ALenum *format,
 
         mem_fclose(instream);
         mem_fclose(outstream);
-
-        S_SetMusicLumpFormatStr("MUS (FluidSynth)");
+        music_format = "MUS (FluidSynth)";
     }
 
     if (result != FLUID_OK)
@@ -486,6 +486,11 @@ static void I_FL_BindVariables(void)
     BIND_BOOL_MIDI(mus_reverb, false, "FluidSynth reverb");
 }
 
+static const char *I_FL_MusicFormat(void)
+{
+    return music_format;
+}
+
 stream_module_t stream_fl_module =
 {
     I_FL_InitStream,
@@ -496,4 +501,5 @@ stream_module_t stream_fl_module =
     I_FL_ShutdownStream,
     I_FL_DeviceList,
     I_FL_BindVariables,
+    I_FL_MusicFormat,
 };
diff --git a/src/i_midimusic.c b/src/i_midimusic.c
index 3b8e1d6b..7f0a7ae9 100644
--- a/src/i_midimusic.c
+++ b/src/i_midimusic.c
@@ -31,7 +31,6 @@
 #include "midifallback.h"
 #include "midifile.h"
 #include "mus2mid.h"
-#include "s_sound.h"
 
 static SDL_Thread *player_thread_handle;
 static SDL_mutex *music_lock;
@@ -1133,8 +1132,6 @@ static boolean RegisterSong(void)
     if (IsMid(song.lump_data, song.lump_length))
     {
         song.file = MIDI_LoadFile(song.lump_data, song.lump_length);
-
-        S_SetMusicLumpFormatStr("MIDI (Native MIDI)");
     }
     else
     {
@@ -1159,8 +1156,6 @@ static boolean RegisterSong(void)
 
         mem_fclose(instream);
         mem_fclose(outstream);
-
-        S_SetMusicLumpFormatStr("MUS (Native MIDI)");
     }
 
     if (song.file == NULL)
@@ -1417,6 +1412,8 @@ static void I_MID_ResumeSong(void *handle)
     SDL_UnlockMutex(music_lock);
 }
 
+static const char *music_format = "Unknown";
+
 static void *I_MID_RegisterSong(void *data, int len)
 {
     if (!music_initialized)
@@ -1424,7 +1421,15 @@ static void *I_MID_RegisterSong(void *data, int len)
         return NULL;
     }
 
-    if (!IsMid(data, len) && !IsMus(data, len))
+    if (IsMid(data, len))
+    {
+        music_format = "MIDI (Native)";
+    }
+    else if (IsMus(data, len))
+    {
+        music_format = "MUS (Native)";
+    }
+    else
     {
         return NULL;
     }
@@ -1498,6 +1503,11 @@ static midiplayertype_t I_MID_MidiPlayerType(void)
     return midiplayer_native;
 }
 
+static const char *I_MID_MusicFormat(void)
+{
+    return music_format;
+}
+
 static void I_MID_BindVariables(void)
 {
     BIND_NUM_MIDI(midi_complevel, COMP_STANDARD, 0, COMP_NUM - 1,
@@ -1525,4 +1535,5 @@ music_module_t music_mid_module =
     I_MID_DeviceList,
     I_MID_BindVariables,
     I_MID_MidiPlayerType,
+    I_MID_MusicFormat,
 };
diff --git a/src/i_oalmusic.c b/src/i_oalmusic.c
index 1c848c24..a52548f8 100644
--- a/src/i_oalmusic.c
+++ b/src/i_oalmusic.c
@@ -486,6 +486,15 @@ static void I_OAL_BindVariables(void)
     }
 }
 
+static const char *I_OAL_MusicFormat(void)
+{
+    if (active_module)
+    {
+        return active_module->I_MusicFormat();
+    }
+    return "None";
+}
+
 music_module_t music_oal_module =
 {
     I_OAL_InitMusic,
@@ -500,4 +509,5 @@ music_module_t music_oal_module =
     I_OAL_DeviceList,
     I_OAL_BindVariables,
     I_OAL_MidiPlayerType,
+    I_OAL_MusicFormat,
 };
diff --git a/src/i_oalstream.h b/src/i_oalstream.h
index 4e3d9a59..07204ce4 100644
--- a/src/i_oalstream.h
+++ b/src/i_oalstream.h
@@ -32,6 +32,7 @@ typedef struct
     void (*I_ShutdownStream)(void);
     const char **(*I_DeviceList)(void);
     void (*BindVariables)(void);
+    const char *(*I_MusicFormat)(void);
 } stream_module_t;
 
 extern stream_module_t stream_opl_module;
diff --git a/src/i_oplmusic.c b/src/i_oplmusic.c
index 47854a73..fa32635a 100644
--- a/src/i_oplmusic.c
+++ b/src/i_oplmusic.c
@@ -1518,6 +1518,8 @@ static boolean I_OPL_InitStream(int device)
 
 static midi_file_t *midifile;
 
+static const char *music_format = "Unknown";
+
 static boolean I_OPL_OpenStream(void *data, ALsizei size, ALenum *format,
                                 ALsizei *freq, ALsizei *frame_size)
 {
@@ -1535,8 +1537,7 @@ static boolean I_OPL_OpenStream(void *data, ALsizei size, ALenum *format,
     if (IsMid(data, size) /* && size < MAXMIDLENGTH */)
     {
         midifile = MIDI_LoadFile(data, size);
-
-        S_SetMusicLumpFormatStr("MIDI (OPL)");
+        music_format = "MIDI (OPL)";
     }
     else
     {
@@ -1561,8 +1562,7 @@ static boolean I_OPL_OpenStream(void *data, ALsizei size, ALenum *format,
 
         mem_fclose(instream);
         mem_fclose(outstream);
-
-        S_SetMusicLumpFormatStr("MUS (OPL)");
+        music_format = "MUS (OPL)";
     }
 
     if (midifile == NULL)
@@ -1690,6 +1690,11 @@ static const char **I_OPL_DeviceList(void)
     return devices;
 }
 
+static const char *I_OPL_MusicFormat(void)
+{
+    return music_format;
+}
+
 static void I_OPL_BindVariables(void)
 {
     BIND_NUM_MIDI(num_opl_chips, 1, 1, OPL_MAX_CHIPS,
@@ -1708,4 +1713,5 @@ stream_module_t stream_opl_module =
     I_OPL_ShutdownStream,
     I_OPL_DeviceList,
     I_OPL_BindVariables,
+    I_OPL_MusicFormat,
 };
diff --git a/src/i_sndfile.c b/src/i_sndfile.c
index 30069dd5..b646a191 100644
--- a/src/i_sndfile.c
+++ b/src/i_sndfile.c
@@ -678,20 +678,6 @@ static boolean I_SND_OpenStream(void *data, ALsizei size, ALenum *format,
     *freq = stream.sfinfo.samplerate;
     *frame_size = stream.frame_size;
 
-    int count;
-    sf_command(NULL, SFC_GET_SIMPLE_FORMAT_COUNT, &count, sizeof(count));
-
-    for (int k = 0; k < count; k++)
-    {
-      SF_FORMAT_INFO format_info = {.format = k};
-      sf_command(NULL, SFC_GET_SIMPLE_FORMAT, &format_info, sizeof(format_info));
-      if (format_info.format == stream.sfinfo.format)
-      {
-        S_SetMusicLumpFormatStr(format_info.name);
-        break;
-      }
-    }
-
     return true;
 }
 
@@ -753,6 +739,28 @@ static void I_SND_BindVariables(void)
     ;
 }
 
+static const char *I_SND_MusicFormat(void)
+{
+    static SF_FORMAT_INFO format_info;
+
+    int count;
+    sf_command(NULL, SFC_GET_SIMPLE_FORMAT_COUNT, &count, sizeof(count));
+
+    for (int i = 0; i < count; i++)
+    {
+        format_info.format = i;
+        sf_command(NULL, SFC_GET_SIMPLE_FORMAT, &format_info,
+                   sizeof(format_info));
+        if (format_info.format == stream.sfinfo.format)
+        {
+            return format_info.name;
+            break;
+        }
+    }
+
+    return "Unknown";
+}
+
 stream_module_t stream_snd_module =
 {
     I_SND_InitStream,
@@ -763,4 +771,5 @@ stream_module_t stream_snd_module =
     I_SND_ShutdownStream,
     I_SND_DeviceList,
     I_SND_BindVariables,
+    I_SND_MusicFormat,
 };
diff --git a/src/i_sound.c b/src/i_sound.c
index 551699ee..5cfac311 100644
--- a/src/i_sound.c
+++ b/src/i_sound.c
@@ -666,8 +666,6 @@ boolean IsMus(byte *mem, int len)
 
 void *I_RegisterSong(void *data, int size)
 {
-    S_SetMusicLumpFormatStr("unknown");
-
     for (int i = 0; i < arrlen(music_modules); ++i)
     {
         void *result = music_modules[i]->I_RegisterSong(data, size);
@@ -728,6 +726,15 @@ const char **I_DeviceList(void)
     return devices;
 }
 
+const char *I_MusicFormat(void)
+{
+    if (active_module)
+    {
+        return active_module->I_MusicFormat();
+    }
+    return "Unknown";
+}
+
 void I_BindSoundVariables(void)
 {
     M_BindNum("sfx_volume", &snd_SfxVolume, NULL, 8, 0, 15, ss_none, wad_no,
diff --git a/src/i_sound.h b/src/i_sound.h
index 49e92b7f..b1de0da5 100644
--- a/src/i_sound.h
+++ b/src/i_sound.h
@@ -167,6 +167,7 @@ typedef struct
     const char **(*I_DeviceList)(void);
     void (*I_BindVariables)(void);
     midiplayertype_t (*I_MidiPlayerType)(void);
+    const char *(*I_MusicFormat)(void);
 } music_module_t;
 
 extern music_module_t music_oal_module;
@@ -207,6 +208,8 @@ boolean IsMid(byte *mem, int len);
 // Determine whether memory block is a .mus file
 boolean IsMus(byte *mem, int len);
 
+const char *I_MusicFormat(void);
+
 void I_BindSoundVariables(void);
 
 #endif
diff --git a/src/i_xmp.c b/src/i_xmp.c
index a9499a4d..199bbe9a 100644
--- a/src/i_xmp.c
+++ b/src/i_xmp.c
@@ -20,7 +20,6 @@
 #include "i_oalstream.h"
 #include "i_printf.h"
 #include "i_sound.h"
-#include "s_sound.h"
 
 static xmp_context context;
 
@@ -96,10 +95,6 @@ static boolean I_XMP_OpenStream(void *data, ALsizei size, ALenum *format,
     *freq = SND_SAMPLERATE;
     *frame_size = 2 * sizeof(short);
 
-    struct xmp_module_info info;
-    xmp_get_module_info(context, &info);
-    S_SetMusicLumpFormatStr(info.mod->type);
-
     return true;
 }
 
@@ -160,6 +155,18 @@ static void I_XMP_BindVariables(void)
     ;
 }
 
+static const char *I_XMP_MusicFormat(void)
+{
+    if (!context)
+    {
+        return "Unknown";
+    }
+
+    static struct xmp_module_info info;
+    xmp_get_module_info(context, &info);
+    return info.mod->type;
+}
+
 stream_module_t stream_xmp_module =
 {
     I_XMP_InitStream,
@@ -170,4 +177,5 @@ stream_module_t stream_xmp_module =
     I_XMP_ShutdownStream,
     I_XMP_DeviceList,
     I_XMP_BindVariables,
+    I_XMP_MusicFormat,
 };
diff --git a/src/s_sound.c b/src/s_sound.c
index d1a9c73a..69e0cf42 100644
--- a/src/s_sound.c
+++ b/src/s_sound.c
@@ -634,16 +634,6 @@ void S_SetSfxVolume(int volume)
 }
 
 static int current_musicnum = -1;
-static char music_lump_format_str[40];
-
-void S_SetMusicLumpFormatStr(const char *str)
-{
-    if (str)
-    {
-        M_StringCopy(music_lump_format_str, str,
-                     sizeof(music_lump_format_str));
-    }
-}
 
 void S_ChangeMusic(int musicnum, int looping)
 {
@@ -695,7 +685,7 @@ void S_ChangeMusic(int musicnum, int looping)
     I_Printf(VB_DEBUG, "S_ChangeMusic: %.8s (%s), %s",
              lumpinfo[music->lumpnum].name,
              W_WadNameForLump(music->lumpnum),
-             music_lump_format_str);
+             I_MusicFormat());
 
     mus_playing = music;
 
@@ -744,7 +734,7 @@ void S_ChangeMusInfoMusic(int lumpnum, int looping)
     I_Printf(VB_DEBUG, "S_ChangeMusInfoMusic: %.8s (%s), %s",
              lumpinfo[music->lumpnum].name,
              W_WadNameForLump(music->lumpnum),
-             music_lump_format_str);
+             I_MusicFormat());
 
     mus_playing = music;
 
diff --git a/src/s_sound.h b/src/s_sound.h
index 3393c201..167fa676 100644
--- a/src/s_sound.h
+++ b/src/s_sound.h
@@ -97,8 +97,6 @@ void S_StartMusic(int music_id);
 void S_ChangeMusic(int music_id, int looping);
 void S_ChangeMusInfoMusic(int lumpnum, int looping);
 
-void S_SetMusicLumpFormatStr(const char *str);
-
 // Stops the music fer sure.
 void S_StopMusic(void);
 
```
</details>